src/mito2/src/memtable/time_series.rs

```diff
@@ -237,39 +247,44 @@ impl Series {
         self.freeze(region_metadata);
 
         let mut frozen = self.frozen.clone();
-        let values = if frozen.len() == 1 {
+        let values = if frozen.is_empty() {
+            Values::new_empty(region_metadata)
+        } else if frozen.len() == 1 {
             frozen.pop().unwrap()
         } else {
             // TODO(hl): We should keep track of min/max timestamps for each values and avoid
             // cloning and sorting when values do not overlap with each other.
 
-            let total_len: usize = frozen.iter().map(|v| v.timestamp.len()).sum();
-            let mut builder = ValueBuilder::new(region_metadata, total_len);
-
-            for v in frozen {
-                let len = v.timestamp.len();
-                builder
-                    .timestamp
-                    .extend_slice_of(&*v.timestamp, 0, len)
-                    .context(CompactValuesSnafu)?;
-                builder
-                    .sequence
-                    .extend_slice_of(&*v.sequence, 0, len)
-                    .context(CompactValuesSnafu)?;
-
-                builder
-                    .op_type
-                    .extend_slice_of(&*v.op_type, 0, len)
-                    .context(CompactValuesSnafu)?;
-
-                for (idx, f) in v.fields.iter().enumerate() {
-                    builder.fields[idx]
-                        .extend_slice_of(&**f, 0, len)
-                        .context(CompactValuesSnafu)?;
-                }
+            let column_size = frozen.get(0).unwrap().fields.len() + 3;
+
+            if cfg!(debug_assertions) {
+                debug_assert!(frozen
+                    .iter()
+                    .zip(frozen.iter().skip(1))
+                    .all(|(prev, next)| { prev.fields.len() == next.fields.len() }));
             }
 
-            let values = Values::from(builder);
+            let mut arrays = frozen.iter().map(|v| v.iter_columns()).collect::<Vec<_>>();
+
+            let cols = (0..column_size)
+                .map(|_| {
+                    arrays
+                        .iter_mut()
+                        .map(|n| n.next().unwrap()) // safety: we've checked the length for every array
+                        .collect::<Vec<_>>()
+                })
+                .collect::<Vec<_>>();
+
+            let concatenated = cols
+                .into_iter()
+                .map(|col| {
+                    arrow::compute::concat(&col.iter().map(|a| a.as_ref()).collect::<Vec<_>>())
+                })
+                .collect::<std::result::Result<Vec<_>, _>>()
+                .context(ComputeArrowSnafu)?;
```

Comment:
Could we return a `Vec<ArrayRef>` in `iter_columns()`? Maybe rename it to `columns()`
```rust
let mut arrays_vec = frozen.iter().map(|v| v.columns()).collect::<Vec<_>>();
let concatenated = (0..column_size)
    .map(|i| {
        // safety: every `Values` has same column num.
        let to_concat = arrays_vec.iter().map(|arrays| arrays[i].as_ref()).collect::<Vec<_>>();
        arrow::compute::concat(&to_concat)
    })
    .collect::<std::result::Result<Vec<_>, _>>()
    .context(ComputeArrowSnafu)?;
```